### PR TITLE
Counts with zeros

### LINKED ISSUE #6937: Pad results with zeros for unmeasured states
### What is the expected enhancement?

After executing a job on qiskit, the typically procedure to get the measurement data from the quantum computer is to call ```get_counts()``` like so...
```
job = execute(qc, backend, shots=100)
result = job.result()
counts = result.get_counts()
```

If the final state is, for example, $| 01 \rangle$ then '01' would be the only key in counts. For some analyses, it would be very useful if the returned counts dictionary had 0s for all the states that were not measured. Maybe something like
```counts = results.get_counts(pad=True)```, and counts would then be ```counts = {'00':0, '01':100, '10':0, '11':0}```

### LINKED ISSUE #7464: KeyError while printing COUNTS of particular qubit
### Environment

- **Qiskit Terra version**: 
- **Python version**: 
- **Operating system**: 

### What is happening?

I apply X-gate to qubit-0
Then I measured it.
I simulate it on qasm_simulator with shots=1024
I plotted the histogram.
I printed counts

So, when I print  counts( ["1"] ), it gave me 1024
BUT when I printed counts counts( ["0"] ), it gave me "KEYERROR" instead of 0.

Kindly resolve this issue.

### How can we reproduce the issue?

I apply X-gate to qubit-0
Then I measured it.
I simulate it on qasm_simulator with shots=1024
I plotted the histogram.
I printed counts

So, when I print  counts( ["1"] ), it gave me 1024
BUT when I printed counts counts( ["0"] ), it gave me "KEYERROR" instead of 0.

Kindly resolve this issue.

### What should happen?

I apply X-gate to qubit-0
Then I measured it.
I simulate it on qasm_simulator with shots=1024
I plotted the histogram.
I printed counts

So, when I print  counts( ["1"] ), it gave me 1024
BUT when I printed counts counts( ["0"] ), it gave me "KEYERROR" instead of 0.

Kindly resolve this issue.

### Any suggestions?

I apply X-gate to qubit-0
Then I measured it.
I simulate it on qasm_simulator with shots=1024
I plotted the histogram.
I printed counts

So, when I print  counts( ["1"] ), it gave me 1024
BUT when I printed counts counts( ["0"] ), it gave me "KEYERROR" instead of 0.

Kindly resolve this issue.

### PR DESCRIPTION
**Pull request by 1ucian0**
Inspired by https://quantumcomputing.stackexchange.com/questions/14029/return-outputs-with-zero-counts

Fixes #6937
Fixes #7464

~This PR adds a zero-default when key not found and extends `qiskit.result.Counts` and `qiskit.result.Result.get_counts` with `include_zeros` (bool, default `False`) to extend the key with zero values.~ 

This PR "mocks" a complete Count dict by returning 0 when the key is missed, considering the possible values of a particular memory size.

### CODE REVIEW

File: qiskit/result/counts.py [hunk 1]
```diff
@@ -180,3 +185,10 @@ def int_outcomes(self):
                 int_key = int(bitstring.replace(" ", ""), 2)
                 out_dict[int_key] = value
             return out_dict
+
+    def __getitem__(self, key):
```
Comment (mtreinish) [inline qiskit/result/counts.py:189]: See the comment on L21 we shouldn't be doing this here it will lead to weird behavior. If we want to do this (which I personally disagree with) we should change the parent here to be defaultdict.

But that being said the purpose of this is not clear to me. The count class is a dict subclass that contains entries for each measurement outcome from an execution. The set of things outside that is infinite and I think a keyerror is the correct response. This also enables incorrect behavior such as:
`Counts()['My Incorrect Answer']` which will return 0.

File: qiskit/result/counts.py [hunk 2]
```diff
@@ -54,38 +54,43 @@ def __init__(self, data, time_taken=None, creg_sizes=None,
...
-            self.hex_raw = {}
-            bin_data = {}
-        else:
+
+        self.int_raw = {}
+        self.hex_raw = {}
+        bin_data = {}
+
+        if include_zeros:
+            for slot_int in range(memory_slots):
```
Comment (mtreinish) [inline qiskit/result/counts.py:71]: You should have a check that memory_slots is actually defined here. If you do `range(None)` this will fail.

File: qiskit/result/counts.py [hunk 3]
```diff
@@ -54,38 +54,43 @@ def __init__(self, data, time_taken=None, creg_sizes=None,
                 ``[('c_reg', 2), ('my_creg', 4)]``.
             memory_slots (int): The number of total ``memory_slots`` in the
                 experiment.
+            include_zeros (bool): If `True`, it includes the counts in zero. Default `False`.
```
Comment (mtreinish) [inline qiskit/result/counts.py:57]: This needs a clearer explanation of what this means, whether there is a 0 or not in the `data` dict depends on what's generating the resutls. IQX might not include every possible bit string but other backends could. I might also rename it to something like `implicit_zeros` to be more descriptive, when I read `include_zeros` I think this means by default we strip 0 values which isn't the case here.

File: qiskit/result/counts.py [hunk 4]
```diff
@@ -54,38 +54,43 @@ def __init__(self, data, time_taken=None, creg_sizes=None,
...
+                bin_data[postprocess._separate_bitstring(slot_bin, creg_sizes)] = 0
+
+        if data:
             first_key = next(iter(data.keys()))
             if isinstance(first_key, int):
-                self.int_raw = data
-                self.hex_raw = {
-                    hex(key): value for key, value in self.int_raw.items()}
+                self.int_raw.update(data)
+                self.hex_raw.update({hex(key): value for key, value in self.int_raw.items()})
```
Comment (mtreinish) [inline qiskit/result/counts.py:81]: Can you not make unrelated whitespace changes here, it makes this harder to review. It took me a while to realize you were not changing the functionality of this line or any of the other ones.

File: qiskit/result/result.py [hunk 5]
```diff
@@ -231,20 +231,20 @@ def get_memory(self, experiment=None):
...
             raise QiskitError('No memory for experiment "{}".'.format(experiment))
 
-    def get_counts(self, experiment=None):
+    def get_counts(self, experiment=None, include_zeros=False):
         """Get the histogram data of an experiment.
 
         Args:
             experiment (str or QuantumCircuit or Schedule or int or None): the index of the
                 experiment, as specified by ``get_data()``.
+            include_zeros (bool): If `True`, it includes the counts in zero. Default `False`.
```
Comment (mtreinish) [inline qiskit/result/result.py:240]: Same comment here as in the class docstring, this needs a better description of how it works.

**Review by mtreinish — CHANGES_REQUESTED** [review]
I disagree with the idea of adding a default value for all invalid keys. To me that isn't the correct direction for this class and is the reason we didn't subclass defaultdict when it was added (also the implementation is incorrect and error prone).

For the other half of this PR though adding an option for implicitly adding a 0 result for other potential values for things in the number of memory slots makes sense to me. Some comments inline on that part.

File: qiskit/result/counts.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/result/counts.py]: changed in 07b89744f

File: qiskit/result/counts.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/result/counts.py]: removed in 9c976307e26d49431c619936cffae76606c1b5c2

File: qiskit/result/counts.py (hunk 2 above)
Comment (1ucian0) [reply qiskit/result/counts.py]: Fixed in 25fbd24

File: qiskit/result/result.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/result/result.py:240]: I think I improved the description in 3f74ff3b4

**Comment by 1ucian0** [comment]
> I disagree with the idea of adding a default value for all invalid keys.

Fair enough. Ill remove that part and we can discuss it later. 

I addressed the rest of the comments.

File: qiskit/result/postprocess.py [hunk 6]
```diff
@@ -60,7 +60,7 @@ def format_counts_memory(shot_memory, header=None):
             specifying the number of total memory_slots in the experiment.
 
     Returns:
-        dict: a formatted memory
+        str: a formatted memory
```
Comment (1ucian0) [inline qiskit/result/postprocess.py:63]: Unrelated fix. I think it returns a `str` in every case.

File: qiskit/result/counts.py (hunk 4 above)
Comment (1ucian0) [reply qiskit/result/counts.py]: fixed in d08b384

**Review by kdk — COMMENTED** [review]
With a reasonable `memory_size` this will grow the size of the result data and the time to build it considerably (to at least 256M at 23 qubits). Instead of building the full dictionary, and to address @mtreinish 's comment, maybe we can validate the key before returning `0`? Otherwise, maybe limit the max `memory_size`?

**Comment by 1ucian0** [comment]
> ..., maybe we can validate the key before returning `0`? Otherwise, maybe limit the max `memory_size`?

followed that way.

**Comment by ajavadia** [comment]
Better be very careful that this does not build all the keys (even valid ones), as that grows exponentially. This is why a dict was used in the first place. The limiting factor should always be number of shots not the possible keys.

Also I don't know why we need to do this. The answer in that stackexchange link seems perfectly reasonable.

**Comment by 1ucian0** [comment]
> Better be very careful that this does not build all the keys (even valid ones), as that grows exponentially. This is why a dict was used in the first place. The limiting factor should always be number of shots not the possible keys.

The current code does not create all the keys but behaves more like a default dict up-to the memory size.

File: qiskit/result/counts.py [hunk 7]
```diff
@@ -103,18 +103,38 @@ def __init__(self, data, time_taken=None, creg_sizes=None, memory_slots=None):
...
+            return super().__getitem__(key)
+        except KeyError as keyerror:
+            if isinstance(key, int):
+                key = hex(key)
+            key = key.replace(" ", "")
+            if not self.bitstring_regex.search(key):
+                key = postprocess.format_counts_memory(key, {"memory_slots": self.memory_slots})
+            if self.memory_slots == len(key):
+                return 0
+            raise KeyError from keyerror
```
Comment (mtreinish) [inline qiskit/result/counts.py:136]: So `__getitem__` was in the comment at the top for a reason, updating dunder methods for a dict subclass is not a good idea and  is explicitly what we should not be doing. Specifically here `get()` will  not have this extra logic now. Basically if you do `get('011', 24)` where `011` is not in the list it will return 24 wihle if you do counts['011'] will return 0. You don't want this to be mismatched like that because it will cause confusing bugs down the road.

The only way to make this work reliably is to change `Counts` to be a `Mapping` implementation and not a dict subclass.

**Review by mtreinish — COMMENTED** [review]
See inline but there is a reason there is a comment at the top about a dict subclass. To make this work consistently as your intending here we'll have to change how the class is structured so it's an implementation of `Mapping` instead of a dict subclass. TBH, I'm still not 100% convinced this is such a useful feature especially when weighing it against the amount of changes we need to make here for it to work.

File: qiskit/result/counts.py (hunk 7 above)
Comment (1ucian0) [reply qiskit/result/counts.py]: done in 92f02a3c8 . Thanks!

**Comment by levbishop** [comment]
I'm not super familiar with this stuff, but I think it might be easier to derive from `UserDict` (which itself is a `Mapping`) and override `__missing__` rather than `__getitem__`.

**Comment by 1ucian0** [comment]
> I'm not super familiar with this stuff, but I think it might be easier to derive from `UserDict` (which itself is a `Mapping`) and override `__missing__` rather than `__getitem__`.

Good one! I'm adding that in bfe9b0d223bb1d86317cac5f92d76dd5d457cd85

File: qiskit/algorithms/amplitude_estimators/ae.py [hunk 8]
```diff
@@ -169,7 +170,7 @@ def evaluate_measurements(
                 y measurements with respective probabilities, in this order.
         """
         # compute grid sample and measurement dicts
-        if isinstance(circuit_results, dict):
+        if isinstance(circuit_results, (dict, Mapping)):
```
Comment (levbishop) [inline qiskit/algorithms/amplitude_estimators/ae.py:173]:
```suggestion
        if isinstance(circuit_results, Mapping):
```
because goose types as a mapping (and other instances below).

File: qiskit/result/counts.py [hunk 9]
```diff
@@ -183,3 +195,20 @@ def int_outcomes(self):
     def _remove_space_underscore(bitstring):
         """Removes all spaces and underscores from bitstring"""
         return int(bitstring.replace(" ", "").replace("_", ""), 2)
+
+    def __iter__(self):
+        return iter(self.bin_data)
+
+    def __len__(self):
+        return len(self.bin_data)
```
Comment (levbishop) [inline qiskit/result/counts.py:203]:
```suggestion
```
These methods are provides by `UserDict`

**Comment by levbishop** [comment]
This now looks like what I had in mind. @mtreinish do you agree with this approach?

File: qiskit/result/counts.py [hunk 10]
```diff
@@ -13,18 +13,15 @@
...
 
-# NOTE: A dict subclass should not overload any dunder methods like __getitem__
-# this can cause unexpected behavior and issues as the cPython dict
-# implementation has many standard methods in C for performance and the dunder
-# methods are not always used as expected. For example, update() doesn't call
-# __setitem__ so overloading __setitem__ would not always provide the expected
-# result
-class Counts(dict):
+class Counts(UserDict, dict):
+    # TODO Remove dict subclassing when ignis get removed, since ignis check for the type
```
Comment (1ucian0) [inline qiskit/result/counts.py:23]:
```suggestion
    # TODO Remove dict subclassing when ignis get removed, since ignis `isinstance(.., dict)`
```

**Comment by 1ucian0** [comment]
With #7464 closed, this PR can be closed too.
